### ISSUE TITLE
Make implicit references to constructors zero-length [phab:D1856]

### ISSUE THREAD

**Issue by kythebot** [google/kythe#2297]
_Created by **zarko** at 2017-08-10 22:10:15_: 

The indexer may emit surprising references to constructors
in common places like return statements (where a move constructor
may be called). We'll leave these in the callgraph (via ref/call),
but we'll mark the ref anchor sources as implicit by making them
zero-length.

This diff also includes a fix for ranges on ctors in new-expressions,
which were one character too short.

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2017-08-10 22:26:14_:

Move this to an internal-linkage function? Should we check that there is exactly one child and that it is a `MaterializeTemporaryExpr`?

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2017-08-10 22:26:14_:

Use `_`? I was running into some minor issues with the 'g' in `Ignore`

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2017-08-10 22:26:14_:

I'm worried that the heuristic isn't quite correct in this case, but the test isn't catching that? I could easily be mistaken, though.

```
|-FunctionDecl 0x25ae5c0 <line:26:1, col:24> col:3 used g '_ (void)'
| `-CompoundStmt 0x25ae808 <col:7, col:24>
|   `-ReturnStmt 0x25ae7f0 <col:9, col:21>
|     `-CXXConstructExpr 0x25ae7b8 <col:16, col:21> '_':'struct S' 'void (struct S &&)' elidable
|       `-MaterializeTemporaryExpr 0x25ae7a0 <col:16, col:21> 'struct S' xvalue
|         `-CXXTemporaryObjectExpr 0x25ae760 <col:16, col:21> 'struct S' 'void (struct S &&)'
|           `-MaterializeTemporaryExpr 0x25ae748 <col:18, col:20> 'struct S' xvalue
|             `-CXXTemporaryObjectExpr 0x25ae6c8 <col:18, col:20> 'struct S' 'void (void)'
```

In particular the inner-most CXXTemporaryObjectExpr{MaterializeTemporarExpr} is an explicit invocation of `S(S())` which we're treating as implicit?

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2017-08-10 22:30:19_:

Also, this test case should be triggering the heuristic (incorrectly?)

```
`-FunctionDecl 0x25b0a20 <line:45:1, col:29> col:4 k '_ *(void)'
  `-CompoundStmt 0x25b0c68 <col:8, col:29>
    `-ReturnStmt 0x25b0c50 <col:10, col:26>
      `-CXXNewExpr 0x25b0bf8 <col:17, col:26> 'struct S *' Function 0x25aec38 'operator new' 'void *(unsigned long)'
        `-CXXConstructExpr 0x25b0bc0 <col:21, col:26> 'struct S' 'void (struct S &&)' elidable
          `-MaterializeTemporaryExpr 0x25b0ba8 <col:23, col:25> 'struct S' xvalue
            `-CXXTemporaryObjectExpr 0x25b0b18 <col:23, col:25> 'struct S' 'void (void)'
```

**Comment by kythe** [comment]
_**shahms** commented inline (context lost during migration) at 2017-08-10 22:36:36_:

ah, yeah, need a test here and above:

//- #0S ref SCtorMove

**Comment by kythe** [comment]
_**zarko** wrote at 2017-08-11 18:10:35_:

  - Switch to a different heuristic

**Comment by kythe** [comment]
_**shahms** wrote at 2017-08-11 18:14:36_:

Thanks Luke!

**Comment by kythe** [comment]
_**shahms** accepted this revision at 2017-08-11 18:14:36_

**Comment by kythe** [comment]
_**zarko** closed via commit 5a76927f2a30a29232185feb005f806780d4d14d at 2017-08-11 19:48:23_
